File: dsa/python/binary_search/binary_search.py
"""
Binary search
"""

def binary_search(arr, value):
    """
    Binary search algorithm
    we could simply use the built-in list.index() method
    but that would be cheating
    return search_list.index(value)

    search_list: list to search
    value: value to search for
    return: index of value in search_list
    """

    lf, rg = 0, len(arr) - 1

    while lf <= rg:
        mid = (rg + lf) // 2

        if arr[mid] < value:
            lf = mid + 1
        elif arr[mid] > value:
            rg = mid - 1
        else:
            return mid
    else:
        raise ValueError("value not in array")

File: dsa/go/binary_search/binary_search.go
package binarysearch

// SearchInts searches for a value in a sorted array of ints
func SearchInts(arr []int, value int) int {
	lf, mid, rg := 0, 0, len(arr)-1

	for lf <= rg {
		mid = (rg + lf) / 2

		if arr[mid] < value {
			lf = mid + 1
		} else if arr[mid] > value {
			rg = mid - 1
		} else {
			return mid
		}
	}
	return -1
}
